Assets/Scripts/Construction/PlayerConstructionHandler.cs

```diff
@@ -0,0 +1,157 @@
+                bool centerSnapX = effectiveGridSize.x % 2 == 1;
+                bool centerSnapZ = effectiveGridSize.y % 2 == 1;
+                Vector3 snappedBuildPoint = SnapToGrid(GetBuildPoint(), centerSnapX, centerSnapZ);
+                currentPlacementPosition = ScanBuildingPlacement(snappedBuildPoint, currentBuildingInfo.GridSize, CurrentConstructionRotation);
+            }
+        }
+
+        private void RotatePlacement(bool reverse=false) 
+        {
+            currentConstructionRotationSteps = (currentConstructionRotationSteps + (reverse ? -1 : 1)) % 4;
+            constructionPreview.SetRotation(CurrentConstructionRotation);
+        }
+
+        private Vector3 GetBuildPoint() 
+        {
+            Ray cameraRay = new Ray(cameraTransform.position, cameraTransform.forward);
+            bool cameraRaySuccess = Physics.Raycast(cameraRay, out RaycastHit cameraRayHit, maxBuildDistance * 2, buildTargetingCollisionMask);
+
+            if (cameraRaySuccess && Vector3.ProjectOnPlane(cameraTransform.position - cameraRayHit.point, Vector3.up).magnitude < maxBuildDistance) {
+                return new Vector3(cameraRayHit.point.x, 0, cameraRayHit.point.z);
+            }
+
+            return Vector3.ProjectOnPlane(cameraTransform.position, Vector3.up) + Vector3.ProjectOnPlane(cameraTransform.forward, Vector3.up) * maxBuildDistance;
+        }
+
+        private Vector3 SnapToGrid(Vector3 vec, bool centerX, bool centerZ) 
+        {
+            vec /= GlobalConstants.GRID_CELL_SIZE;
+            Vector3 newVec = new Vector3(Mathf.Round(vec.x), 0, Mathf.Round(vec.z));
+            newVec *= GlobalConstants.GRID_CELL_SIZE;
+
+            return newVec + new Vector3(centerX ? GlobalConstants.GRID_CELL_SIZE/2.0f : 0, 0, centerZ ? GlobalConstants.GRID_CELL_SIZE/2.0f : 0);
+        }
+
+        // Returns the ultimate placement position of the building
+        private Vector3? ScanBuildingPlacement(Vector3 buildPoint, Vector2Int buildingGridSize, Quaternion buildingRotation) 
+        {
+            Vector3 boxOrigin = new Vector3(buildPoint.x, maxTerrainHeight+1, buildPoint.z);
+            Vector3 boxHalfExtents = new Vector3(buildingGridSize.x * GlobalConstants.GRID_CELL_SIZE * 0.5f, 1, buildingGridSize.y * GlobalConstants.GRID_CELL_SIZE * 0.5f);
+
+            bool success = Physics.BoxCast(boxOrigin, boxHalfExtents, Vector3.down, out RaycastHit result, buildingRotation, maxTerrainHeight*2, buildTargetingCollisionMask);
+            if (success == false) {
+                constructionPreview.gameObject.SetActive(false);
+                return null;
+            }
+            constructionPreview.gameObject.SetActive(true);
+
+            Vector3 finalPlacementPosition = new Vector3(buildPoint.x, result.point.y, buildPoint.z);
+
+            constructionPreview.SetPosition(finalPlacementPosition);
+
+            return finalPlacementPosition;
+        }
```

Comment:
I would suggest changing the name of this function to CalculateBuildingPlacement, also remove both setActive and constructionPreview.SetPosition from this function. We only want to calculate the building placement in here. We can SetActive and preview inside ProcessPlacement as it is a more general function. Let's not make 'Ultimate functions'. Use the nullable vector to determine if we need to change the state.